distributed/client.py

```diff
@@ -1441,7 +1441,10 @@ def wait_for_workers(
         return self.sync(self._wait_for_workers, n_workers, timeout=timeout)
 
     def _heartbeat(self):
-        if self.scheduler_comm:
+        # Don't send heartbeat if scheduler comm or cluster are already closed
+        if (self.scheduler_comm and not self.scheduler_comm.comm.closed()) or (
+            self.cluster and self.cluster.status not in (Status.closed, Status.closing)
+        ):
```

Comment:
I'm slightly confused by this.  If `scheduler_comm.comm.closed()` should we send a heartbeat if the cluster is not closing?

If the client closes but the cluster is still alive then we should probably stop?

Reply:
In particular, I think that this case might be rare but relevant

```python
cluster = Cluster()

client1 = Client(cluster)
client2 = Client(cluster)

client1.close()
```

Reply:
> If scheduler_comm.comm.closed() should we send a heartbeat if the cluster is not closing?

I don't think we want to attempt to send a heartbeat if the `scheduler_comm` is closed. This `if`-statement should include that case. 

> If the client closes but the cluster is still alive then we should probably stop?

This should already be the case. If the client closes then the `scheduler_comm` will also be closed

```python
In [1]: from distributed import LocalCluster, Client

In [2]: cluster = LocalCluster()
   ...:
   ...: client1 = Client(cluster)
   ...: client2 = Client(cluster)
   ...:
   ...: client1.close()

In [3]: client1.scheduler_comm.comm.closed()
Out[3]: True

In [4]: client2.scheduler_comm.comm.closed()
Out[4]: False
```